workflows/seqinspector.nf

```diff
@@ -91,109 +101,178 @@ workflow SEQINSPECTOR {
         ch_versions = ch_versions.mix(FASTQSCREEN_FASTQSCREEN.out.versions.first())
     }
-    // MODULE: Create BWA-MEM2 index of the reference genome
-
+    // MODULE: Create BWA-MEM2 index of the reference genome OR use pre-built index
     if (!("bwamem2_index" in skip_tools)) {
-        ch_reference_fasta  = Channel.fromPath(params.fasta).map { it -> [[id:it.simpleName], it] }.collect()
+        // Always create the reference FASTA channel since it's needed by BWAMEM2_MEM
+        ch_reference_fasta = channel.fromPath(fasta_file, checkIfExists: true).map { file -> tuple([id: file.name], file) }.collect()
 
-        BWAMEM2_INDEX (
-            ch_reference_fasta
-        )
-        ch_bwamem2_index = BWAMEM2_INDEX.out.index
-        ch_versions = ch_versions.mix(BWAMEM2_INDEX.out.versions)
+        if (params.bwa_index) {
+            // Use pre-built index when --bwa_index parameter is provided
+            ch_bwamem2_index = channel.fromPath(params.bwa_index, checkIfExists: true)
+                .map { index_dir -> tuple([id: index_dir.name], index_dir) }
+                .collect()
 
+            // Add debug output to verify the pre-built index is being used
+            ch_bwamem2_index.view { "Using pre-built BWA-MEM2 index: ${it}" }
+        }
+        else {
+            // Build index from reference FASTA when no pre-built index is provided
+            BWAMEM2_INDEX(
+                ch_reference_fasta
+            )
+            ch_bwamem2_index = BWAMEM2_INDEX.out.index
+            ch_versions = ch_versions.mix(BWAMEM2_INDEX.out.versions)
+
+            // Add debug output to verify the index was built
+            ch_bwamem2_index.view { "Built BWA-MEM2 index: ${it}" }
+        }
     }
+
     // MODULE: Align reads with BWA-MEM2
     if (!("bwamem2_mem" in skip_tools)) {
-        BWAMEM2_MEM (
+        BWAMEM2_MEM(
             ch_sample_sized,
             ch_bwamem2_index,
             ch_reference_fasta,
-            params.sort_bam ?: true
+            params.sort_bam ?: true,
         )
         ch_bwamem2_mem = BWAMEM2_MEM.out.bam
         ch_versions = ch_versions.mix(BWAMEM2_MEM.out.versions)
-        ch_bwamem2_mem.view { "BAM: $it" }
+        ch_bwamem2_mem.view { "BAM: ${it}" }
+    }
+    // MODULE: Index BAM files with Samtools
+    if (!("samtools_index" in skip_tools) && !("bwamem2_mem" in skip_tools)) {
+        SAMTOOLS_INDEX(
+            ch_bwamem2_mem
+        )
+        ch_samtools_index = SAMTOOLS_INDEX.out.bai
+        ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions)
+        ch_samtools_index.view { "BAI: ${it}" }
+    }
 
+    // MODULE: Index reference FASTA with Samtools faidx
+    if (!("samtools_faidx" in skip_tools)) {
 
-}
+        // Assume ch_fasta emits tuple(meta, fasta)
+        ch_dummy_fai = channel.value(['dummy_meta', file('empty.fai')])
+
+
+        SAMTOOLS_FAIDX(
+            ch_reference_fasta,
+            ch_dummy_fai,
+            true,
+        )
```

Comment:
I have a feeling this should be done in a separate prepare genome subworkflows prior to this actual workflow, happy to talk with you about some refactoring